### PR TITLE
Adjust GetBulkFromEpoch() - keep order of requested items

### PR DESCRIPTION
**Pull request by andreibancioiu**
Adjusted signature of `Storer.GetBulkFromEpoch()`: return a list of key-value pairs instead of a map. The items are returned in their requested order.

**Important for review :exclamation:**  previously, `GetBulkFromEpoch` automatically performed a union over the returned items - that is, duplicate keys in the input list did not result into duplicate items in the returned map (since it was a map, after all). Now, duplicate keys in the input list **will** result into duplicate items in the returned list. **However**, we can consider this expected behavior. `GetBulkFromEpoch` is only used in the API components in production code - therefore, no protocol change.

Furthermore, adjusted tests to use the `StorerMock` from `testscommon`, where possible. Could not adjust the tests in `epochStart/metachain` - some of them rely more on the current behavior of `process/mock/StorerMock` and `epochStart/mock/StorerMock`.

### CODE REVIEW

File: epochStart/mock/storerMock.go [hunk 1]
```diff
@@ -5,6 +5,8 @@ import (
 	"errors"
 	"fmt"
 	"sync"
+
+	"github.com/ElrondNetwork/elrond-go/storage"
```
Comment (bogdan-rosianu) [inline epochStart/mock/storerMock.go:9]: We shouldn't import the `storage` package which is quite heavy only for the simple struct `KeyValuePair` which only contains `Key` and `Value`. Maybe move `KeyValuePair` to `common` package with a TODO to move it to core

File: storage/keyDataPair.go [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package storage
+
+type KeyValuePair struct {
```
Comment (bogdan-rosianu) [inline storage/keyDataPair.go:3]: please move it elsewhere

File: storage/pruning/pruningStorer.go [hunk 3]
```diff
@@ -523,10 +523,10 @@ func (ps *PruningStorer) GetBulkFromEpoch(keys [][]byte, epoch uint32) (map[stri
 			continue
 		}
 
-		returnMap[string(key)] = res
+		results = append(results, storage.KeyValuePair{Key: key, Value: res})
```
Comment (bogdan-rosianu) [inline storage/pruning/pruningStorer.go:526]: regarding the duplications from the PR description, we cannot have any duplicates this way because the storer will contain unique keys. Therefore, I don't see any case where we can have duplicates

Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go:526]: Previously, slightly-corrupted input (duplicated requested keys) resulted into non-corrupted output (duplicated requested keys removed). Now, duplicated requested keys would result into duplicated response items (even though the `storer` cannot have duplicate keys).

Comment (bogdan-rosianu) [reply storage/pruning/pruningStorer.go:526]: I see. I think that the current code is ok. Our code that requests txs will not call for duplicated, so we're fine. No need to further filter the transactions

File: epochStart/mock/storerMock.go (hunk 1 above)
Comment (andreibancioiu) [reply epochStart/mock/storerMock.go]: Fixed, moved to `common`. Added TODO.

File: storage/keyDataPair.go (hunk 2 above)
Comment (andreibancioiu) [reply storage/keyDataPair.go]: Fixed, moved.

File: node/external/blockAPI/baseBlock.go [hunk 4]
```diff
@@ -162,11 +162,11 @@ func (bap *baseAPIBlockProcessor) getReceiptsFromMiniblock(miniblock *block.Mini
 	log.Debug(fmt.Sprintf("GetBulkFromEpoch took %s", time.Since(start)))
 
 	apiReceipts := make([]*transaction.ApiReceipt, 0)
-	for recHash, recBytes := range marshalizedReceipts {
-		receipt, errUnmarshal := bap.txUnmarshaller.UnmarshalReceipt(recBytes)
+	for _, pair := range marshalizedReceipts {
```
Comment (iulianpascalau) [inline node/external/blockAPI/baseBlock.go:165]: suggestion to rename marshalizedReceipts to marshalledReceipts

File: process/mock/storerMock.go [hunk 5]
```diff
@@ -57,22 +54,7 @@ func (sm *StorerMock) GetFromEpoch(key []byte, _ uint32) ([]byte, error) {
...
-		}
-		retValue[string(key)] = value
-	}
-
-	return retValue, nil
-}
-
-// SearchFirst -
-func (sm *StorerMock) SearchFirst(_ []byte) ([]byte, error) {
+func (sm *StorerMock) GetBulkFromEpoch(keys [][]byte, _ uint32) ([]common.KeyValuePair, error) {
```
Comment (iulianpascalau) [inline process/mock/storerMock.go:57]: why this change on the mock?
Missing comment

File: process/mock/storerMock.go [hunk 6]
```diff
@@ -81,26 +63,23 @@ func (sm *StorerMock) Has(_ []byte) error {
...
 
+// SearchFirst -
+func (sm *StorerMock) SearchFirst(key []byte) ([]byte, error) {
+	return sm.Get(key)
+}
+
 // RemoveFromCurrentEpoch -
 func (sm *StorerMock) RemoveFromCurrentEpoch(key []byte) error {
-	return sm.Remove(key)
+	delete(sm.data, string(key))
```
Comment (iulianpascalau) [inline process/mock/storerMock.go:73]: why this change?

File: process/mock/storerMock.go [hunk 7]
```diff
@@ -81,26 +63,23 @@ func (sm *StorerMock) Has(_ []byte) error {
...
+	return nil
 }
 
 // Remove -
 func (sm *StorerMock) Remove(key []byte) error {
-	sm.mut.Lock()
```
Comment (iulianpascalau) [inline process/mock/storerMock.go:91]: is this ok? This is not safe to be ran concurrently anymore

File: storage/pruning/fullHistoryPruningStorer.go [hunk 8]
```diff
@@ -79,22 +80,22 @@ func (fhps *FullHistoryPruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]
...
 		return nil, err
 	}
 
-	results := make(map[string][]byte)
+	results := make([]common.KeyValuePair, 0, len(keys))
 	for _, key := range keys {
 		dataInCache, found := fhps.cacher.Get(key)
 		if found {
-			results[string(key)] = dataInCache.([]byte)
+			results = append(results, common.KeyValuePair{Key: key, Value: dataInCache.([]byte)})
```
Comment (iulianpascalau) [inline storage/pruning/fullHistoryPruningStorer.go:93]: split in multiple?

File: storage/pruning/fullHistoryPruningStorer.go [hunk 9]
```diff
@@ -79,22 +80,22 @@ func (fhps *FullHistoryPruningStorer) GetFromEpoch(key []byte, epoch uint32) ([]
...
 			continue
 		}
 		data, errGet := persister.Get(key)
 		if errGet == nil && data != nil {
-			results[string(key)] = data
+			results = append(results, common.KeyValuePair{Key: key, Value: data})
```
Comment (iulianpascalau) [inline storage/pruning/fullHistoryPruningStorer.go:98]: split in multiple lines?

File: storage/pruning/pruningStorer.go [hunk 10]
```diff
@@ -506,11 +506,11 @@ func (ps *PruningStorer) GetBulkFromEpoch(keys [][]byte, epoch uint32) (map[stri
 	}
 	defer closePersister()
 
-	returnMap := make(map[string][]byte)
+	results := make([]common.KeyValuePair, 0, len(keys))
 	for _, key := range keys {
 		v, ok := ps.cacher.Get(key)
 		if ok {
-			returnMap[string(key)] = v.([]byte)
+			results = append(results, common.KeyValuePair{Key: key, Value: v.([]byte)})
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:513]: multiple lines?

File: storage/pruning/pruningStorer.go [hunk 11]
```diff
@@ -523,10 +523,10 @@ func (ps *PruningStorer) GetBulkFromEpoch(keys [][]byte, epoch uint32) (map[stri
 			continue
 		}
 
-		returnMap[string(key)] = res
+		results = append(results, common.KeyValuePair{Key: key, Value: res})
```
Comment (iulianpascalau) [inline storage/pruning/pruningStorer.go:526]: multiple lines?

File: storage/storageUnit/storageunit.go [hunk 12]
```diff
@@ -196,9 +196,9 @@ func (u *Unit) GetBulkFromEpoch(keys [][]byte, _ uint32) (map[string][]byte, err
 			)
 			continue
 		}
-		retMap[string(key)] = value
+		results = append(results, common.KeyValuePair{Key: key, Value: value})
```
Comment (iulianpascalau) [inline storage/storageUnit/storageunit.go:199]: multiple lines?

File: testscommon/genericMocks/storerMock.go [hunk 13]
```diff
@@ -32,6 +33,14 @@ func NewStorerMock(name string, currentEpoch uint32) *StorerMock {
 	return sm
 }
 
+func NewStorerMockWithEpoch(currentEpoch uint32) *StorerMock {
```
Comment (iulianpascalau) [inline testscommon/genericMocks/storerMock.go:36]: missing comments

File: testscommon/genericMocks/storerMock.go [hunk 14]
```diff
@@ -32,6 +33,14 @@ func NewStorerMock(name string, currentEpoch uint32) *StorerMock {
...
+	return NewStorerMock("Storage", currentEpoch)
+}
+
+func NewStorerMockWithDefaults() *StorerMock {
```
Comment (iulianpascalau) [inline testscommon/genericMocks/storerMock.go:40]: We will never know what `WithDefaults` means and go back an forth with this file.
Let's clean this file and just rename this to `NewStorerMock` -> the other one is actually `NewStorerMockWithEpoch`, so another rename

File: testscommon/genericMocks/storerMock.go [hunk 15]
```diff
@@ -78,18 +87,18 @@ func (sm *StorerMock) GetFromEpoch(key []byte, epoch uint32) ([]byte, error) {
...
+func (sm *StorerMock) GetBulkFromEpoch(keys [][]byte, epoch uint32) ([]common.KeyValuePair, error) {
 	data := sm.GetEpochData(epoch)
-	result := map[string][]byte{}
+	results := make([]common.KeyValuePair, 0, len(keys))
 
 	for _, key := range keys {
 		value, ok := data.Get(string(key))
 		if ok {
-			result[string(key)] = value.([]byte)
+			results = append(results, common.KeyValuePair{Key: key, Value: value.([]byte)})
```
Comment (iulianpascalau) [inline testscommon/genericMocks/storerMock.go:97]: multiple lines?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

**Review by miiu96 — DISMISSED** [review]

File: common/keyValuePair.go [hunk 16]
```diff
@@ -0,0 +1,8 @@
+package common
+
+// KeyValuePair is a tuple of (key, value)
+// TODO: Move this to elrond-go-core.
+type KeyValuePair struct {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply common/keyValuePair.go:5]: Thanks, missed that. The struct isn't exported, though. One option would be to define an interface for `KeyValueStorage` in `storer/interface.go`. Another option would be to convert `keyValStorage` to an exported plain struct, to always be passed by value (of course, it will still hold pointers to key & value).

Keep as it is? What would be best in this case?

File: epochStart/mock/storerMock.go [hunk 17]
```diff
@@ -52,18 +54,8 @@ func (sm *StorerMock) GetFromEpoch(key []byte, _ uint32) ([]byte, error) {
...
-		value, err := sm.Get(key)
-		if err != nil {
-			continue
-		}
-		retValue[string(key)] = value
-	}
-
-	return retValue, nil
+func (sm *StorerMock) GetBulkFromEpoch(keys [][]byte, _ uint32) ([]common.KeyValuePair, error) {
+	return nil, errors.New("not implemented")
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply epochStart/mock/storerMock.go:58]: Added comment, left not implemented (not used).

File: node/external/blockAPI/baseBlock.go (hunk 4 above)
Comment (andreibancioiu) [reply node/external/blockAPI/baseBlock.go:165]: Renamed.

File: node/external/blockAPI/baseBlock.go [hunk 18]
```diff
@@ -196,15 +196,15 @@ func (bap *baseAPIBlockProcessor) getTxsFromMiniblock(
 
 	start = time.Now()
 	txs := make([]*transaction.ApiTransactionResult, 0)
-	for txHash, txBytes := range marshalizedTxs {
-		tx, errUnmarshalTx := bap.txUnmarshaller.UnmarshalTransaction(txBytes, txType)
+	for _, pair := range marshalizedTxs {
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply node/external/blockAPI/baseBlock.go:199]: Renamed.

File: process/mock/storerMock.go (hunk 5 above)
Comment (andreibancioiu) [reply process/mock/storerMock.go:57]: Added comment, left not implemented (function not used).

File: process/mock/storerMock.go (hunk 6 above)
Comment (andreibancioiu) [reply process/mock/storerMock.go:73]: It was by mistake (removing and reverting the mocks etc.). Reverted :pray:

File: process/mock/storerMock.go (hunk 7 above)
Comment (andreibancioiu) [reply process/mock/storerMock.go:91]: Reverted (copy-pasted another mock by mistake).

File: storage/pruning/fullHistoryPruningStorer.go (hunk 8 above)
Comment (andreibancioiu) [reply storage/pruning/fullHistoryPruningStorer.go:93]: Fixed.

File: storage/pruning/fullHistoryPruningStorer.go (hunk 9 above)
Comment (andreibancioiu) [reply storage/pruning/fullHistoryPruningStorer.go:98]: Fixed.

File: storage/pruning/pruningStorer.go (hunk 10 above)
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go:513]: Fixed.

File: storage/pruning/pruningStorer.go (hunk 11 above)
Comment (andreibancioiu) [reply storage/pruning/pruningStorer.go:526]: Fixed.

File: storage/storageUnit/storageunit.go (hunk 12 above)
Comment (andreibancioiu) [reply storage/storageUnit/storageunit.go:199]: Fixed.

File: testscommon/genericMocks/storerMock.go (hunk 15 above)
Comment (andreibancioiu) [reply testscommon/genericMocks/storerMock.go:97]: Fixed.

File: testscommon/genericMocks/storerMock.go (hunk 13 above)
Comment (andreibancioiu) [reply testscommon/genericMocks/storerMock.go:36]: Added.

File: testscommon/genericMocks/storerMock.go (hunk 14 above)
Comment (andreibancioiu) [reply testscommon/genericMocks/storerMock.go:40]: Fixed.

File: process/mock/storerMock.go (hunk 5 above)
Comment (iulianpascalau) [reply process/mock/storerMock.go:63]: ok

**Review by iulianpascalau — APPROVED** [review]

File: common/keyValuePair.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply common/keyValuePair.go:5]: I think that you should keep this `KeyValuePair` (and move it to core), because the keyValStorage is pretty specialized for trie values, as the struct also contains the function ValueWithoutSuffix(). This way, keyValStorage can "extend" `KeyValuePair`

**Review by bogdan-rosianu — APPROVED** [review]
